src/FastMathFunctions.cpp

```diff
@@ -0,0 +1,1074 @@
+    Expr result = eval_approx(approx, x);
+    result = select(flip_sign, -result, result);
+    result = common_subexpression_elimination(result, true);
+    return result;
+}
+
+Expr fast_cos(const Expr &x_full, ApproximationPrecision precision) {
+    const bool use_sin = precision.optimized_for == ApproximationPrecision::MULPE;
+
+    Type type = x_full.type();
+    Expr x_abs = abs(x_full);
+    // Range reduction to interval [0, pi/2] which corresponds to a quadrant of the circle.
+    Expr scaled = x_abs * make_const(type, TWO_OVER_PI);
+    Expr k_real = floor(scaled);
+    Expr k = cast<int>(k_real);
+    Expr k_mod4 = k % 4;  // Halide mod is always positive!
+    Expr mirror = ((k_mod4 == 1) || (k_mod4 == 3));
+    if (use_sin) {
+        mirror = !mirror;
+    }
+    Expr flip_sign = ((k_mod4 == 1) || (k_mod4 == 2));
+
+    // Reduce the angle modulo pi/2: i.e., to the angle within the quadrant.
+    Expr x = x_abs - k_real * make_const(type, PI_OVER_TWO);
+    Expr pi_over_two_minus_x = make_const(type, PI_OVER_TWO) - x;
+    if (type == Float(32) && precision.optimized_for == ApproximationPrecision::MULPE) {
+        auto [hi, lo] = split_float(PI_OVER_TWO);
+        // TODO(mcourteaux): replace with proper strict_float intrinsic ops.
+        pi_over_two_minus_x = strict_add(strict_sub(make_const(type, hi), x), make_const(type, lo));
+    }
+    x = select(mirror, pi_over_two_minus_x, x);
+
+    Expr result;
+    if (use_sin) {
+        // Approximating cos(x) as sin(pi/2 - x).
+        const Internal::Approximation *approx = Internal::ApproximationTables::best_sin_approximation(precision, type);
+        result = eval_approx(approx, x);
+    } else {
+        const Internal::Approximation *approx = Internal::ApproximationTables::best_cos_approximation(precision, type);
+        result = eval_approx(approx, x);
+    }
+    result = select(flip_sign, -result, result);
+    result = common_subexpression_elimination(result, true);
+    return result;
+}
+
+Expr fast_tan(const Expr &x_full, ApproximationPrecision precision) {
+    Type type = x_full.type();
+
+    // Reduce range to [-pi/2, pi/2]
+    Expr scaled = x_full * make_const(type, ONE_OVER_PI);
+    Expr k_real = round(scaled);
+
+    Expr x = x_full - k_real * make_const(type, PI);
+    if (type == Float(32) && precision.optimized_for == ApproximationPrecision::MULPE) {
+        auto [pi_hi, pi_lo] = split_float(PI);
+        x = strict_sub((x_full - k_real * make_const(type, pi_hi)), (k_real * make_const(type, pi_lo)));
+    }
+
+    // When polynomial: x is assumed to be reduced to [-pi/2, pi/2]!
+    const Internal::Approximation *approx = Internal::ApproximationTables::best_tan_approximation(precision, type);
+
+    Expr abs_x = abs(x);
+    Expr flip = x < make_const(type, 0.0);
+    Expr use_cotan = abs_x > make_const(type, PI / 4.0);
+    Expr pi_over_two_minus_abs_x;
+    if (type == Float(64)) {
```

Comment:
Changing both branches of the if to use strict add or sub fixes the WebAssembly failures. This is due to reassociation in fast math -- as in commenting out `fp_flags.setAllowReassoc` in `CodeGen_LLVM::set_fp_fast_math` makes the tests pass for wasm without this change. Thus I'm guessing strictness is legitimately needed here.

`    if (type == Float(64)) {
        // TODO(mcourteaux): We could do split floats here too.
        pi_over_two_minus_abs_x = strict_sub(make_const(type, PI_OVER_TWO), abs_x);
    } else if (type == Float(32)) {  // We want to do this trick always, because we invert later.
        auto [hi, lo] = split_float(PI_OVER_TWO);
        pi_over_two_minus_abs_x = strict_add(strict_sub(make_const(type, hi), abs_x), make_const(type, lo));
    }
`

Reply:
Maybe LLVM does reassociative rewrites/folding if the root node (+) is non-strict, without checking if the child nodes (-) are also non-strict.

Reply:
That would be very strange and render every LLVM expression in IR non-strict. For example: a binary tree of 5 levels deep of nested strict operations, followed by a non-strict + 1, destroys the whole strict tree? That's at least how I interpret your guess.

Well, I can see that if the non-strict + is directly followed by a strict +, you can tell that the secondary add is not gonna do anything if you constant fold it with the other addition. So you could argue that the "optimization" is correct, because the addition was non-strict, so LLVM decided it would not have any impact. I am not too sure if that reason is sound. However, adding strictness to the second op should be fine either way. We *do* want the add, so making it strict will never hurt.